docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,229 @@
+    pub fn subscribe_for_shutdown(&self) -> DropWatcher;
+
+    /// Applies the specified function to the tenant. Returns error if the tenant is
+    /// not ready to launch the closure.
+    pub fn with_tenant<F, O>(&self, func: F) -> anyhow::Result<O>
+    where
+    F: FnOnce(TenantSyncRef) -> O {}
+
+    /// Applies the specified function to the tenant. Returns error if the tenant is
+    /// not ready to launch the closure.
+    pub async fn with_tenant_async<F, T, O>(&self, func: F) -> anyhow::Result<O>
+    where
+    F: FnOnce(TenantAsyncRef) -> T,
+    T: Future<Output = O> {}
+}
+
+impl Drop for TenantAccessor {
+    // schedule a drop of itself in tokio runtime with necessary joins...
+}
+```
+
+### `TenantAccessor` and states
+
+This is how the accessor behaves when it sees different states:
+
+1. `Infant` - we cannot create a guard around the tenant at this stage. We should start loading some layers and become `Loading` or `Downloading` first and only after creating an accessor.
+2. `Loading` - we cannot run any operations on the tenant until it is either in an `Active` or `Broken` state.
+3. `Broken` - we cannot run any operations on the tenant.
+4. `Downloading` - we cannot run any operations on the tenant until it is either in an `Active` or `Broken` state.
+5. `Active` - the tenant can run any closures. **That means it's the only way to run any closures on tenants**.
+6. `ShuttingDown` - we cannot run any operations on the tenant.
+7. `Shutdown` - we cannot run any operations on the tenant.
+
+### `TenantAccessor` and cancelations
+
+One great question is: what guarantees do we have about tenants inside the closures we're running?
+
+After we introduced the `with_tenant` functions, there are a bunch of closures running. The only way to run a closure is when we are under an `Active` state. What if one closure wants to run under this state, and another decides to shut down the tenant? Do we want to continue executing?
+
+In this specific RFC:
+
+1. We use a `StateKeeper` / `StateSubscriber` / `StateHider` primitive to ensure all closures are gone, and running closures can't tell the difference.
+2. It's proposed to run only non-cancellable short-living closures and check in the code reviews that the closure won't run for a long time. This should be easy to spot. Moreover: _all of the current code does not use long-running functions inside tenants_, even when GC or compaction is run.
+3. To cancel GC and compaction, we could use the `LoopedTaskOperator` / `LoopedTaskController` pair. After they finish iteration, they check whether they should stop, and we wait for this to happen. Also, we can receive commands to pause, continue and continue the loop once.
+
+## `TimelineAccessor`
+
+Works the same as `TenantAccessor`. Problems like [(#2442)](https://github.com/neondatabase/neon/issues/2442) are solved since we have to use `with_timeline` to lock on the timeline, and if GC or compaction is run, nothing will go wrong.
```

Comment:
I'm not sure that it solve the problem. Personally I dont see problem in this case. We either wait when there is no running task, or return en error and expect that the call will be retried. Current approach is to return error. And the issue is caused by absence of retries in test client. 

Waiting can cause timeouts on the other side. If these calls are retried we can get indefinite number of tasks waiting for gc/compaction to terminate. Returning an error quickly allows to release resources and pass this retrying mechanics to control plane which already has decent mechanism for retrying such operations. We need to just return proper error codes so it can be sure that something can be retried.

Both approaches can be implemented with current task_mgr and with things proposed in this RFC

Reply:
I can't understand what retries you mean. We either get access instantly or get an error.